### PR TITLE
[Bug bash] Improve serde logic and log messages for LlamaIndex

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12716?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12716/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12716
```

</p>
</details>

### What changes are proposed in this pull request?

1. Remove the static mapping and use the standard `BaseComponent` class for checking if the object is serializable or not for better maintainability (ref: [comment](https://github.com/mlflow/mlflow/pull/12704#discussion_r1682254152))
2. Avoid showing `API key removed` warning multiple times.
3. Avoid showing `Serialization of the following objects are not supported` warning for `None` (or empty default callback manager) for the Setting.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

**Before PR**
```
2024/07/19 11:03:49 INFO mlflow.llama_index.serialize_objects: API key removed from object serialization. At inference time, the key must be passed as an environment variable or via inference parameters.
2024/07/19 11:03:49 INFO mlflow.llama_index.serialize_objects: API key removed from object serialization. At inference time, the key must be passed as an environment variable or via inference parameters.
2024/07/19 11:03:49 INFO mlflow.llama_index.serialize_objects: Serialization of the following objects are not supported and will not be logged with your model: _callback_manager, _tokenizer, _prompt_helper
```

**After PR**
```
2024/07/19 11:25:31 INFO mlflow.llama_index.serialize_objects: API key(s) will be deducted from the global Settings object during serialization. At inference time, the key(s) must be passed as environment variables.
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/llama_index/serialize_objects.py [hunk 1]
```diff
@@ -53,32 +28,28 @@ def _get_object_import_path(o: object) -> str:
 
 
 def _sanitize_api_key(object_as_dict: Dict[str, str]) -> Dict[str, str]:
-    keys_to_remove = [k for k in object_as_dict.keys() if "api_key" in k.lower()]
+    return {k: v for k, v in object_as_dict.items() if "api_key" not in k.lower()}
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:31]: 👍

File: mlflow/llama_index/serialize_objects.py [hunk 2]
```diff
@@ -159,38 +119,60 @@ def _deserialize_dict_of_objects(path: str) -> Dict[str, any]:
         return output
 
 
-def _serialize_dict_of_objects(dict_of_objects: Dict[str, object], path: str) -> None:
+def serialize_settings(path: str) -> None:
+    """Serialize the global LlamaIndex Settings object to a JSON file at the given path."""
+    from llama_index.core import Settings
+
+    _logger.info(
+        "API key(s) will be deducted from the global Settings object during serialization. "
```
Comment (BenWilson2) [inline mlflow/llama_index/serialize_objects.py:127]:
```suggestion
        "API key(s) will be removed from the global Settings object during serialization to protect against key leakage. "
```

**Review by BenWilson2 — APPROVED** [review]
Love the simplifications here. Thanks @B-Step62 !!

**Comment by B-Step62** [comment]
The doc job failure is irrelevant to this PR. Will address it in the other PR. https://github.com/mlflow/mlflow/pull/12719

**Pull request merged**